plutus-core/plutus-core/src/PlutusCore/Bitwise/Logical.hs

```diff
@@ -0,0 +1,461 @@
+-- editorconfig-checker-disable-file
+
+{-# LANGUAGE OverloadedStrings #-}
+
+-- | Implementations of bitwise logical primops.
+module PlutusCore.Bitwise.Logical (
+  bitwiseLogicalAnd,
+  bitwiseLogicalOr,
+  bitwiseLogicalXor,
+  bitwiseLogicalComplement,
+  readBit,
+  writeBits,
+  replicateByteString
+  ) where
+
+import Control.Exception (Exception, throw, try)
+import Data.Bits qualified as Bits
+import Data.ByteString (ByteString)
+import Data.ByteString qualified as BS
+import Data.ByteString.Internal qualified as BSI
+import Data.Foldable (for_, traverse_)
+import Data.Text (pack)
+import Data.Word (Word64, Word8)
+import Foreign.Marshal.Utils (copyBytes)
+import Foreign.Ptr (Ptr, castPtr, plusPtr)
+import Foreign.Storable (peekByteOff, peekElemOff, pokeByteOff, pokeElemOff)
+import PlutusCore.Builtin (BuiltinResult, emit)
+import PlutusCore.Evaluation.Result (evaluationFailure)
+import System.IO.Unsafe (unsafeDupablePerformIO)
+
+{- Note [Binary bitwise operation implementation and manual specialization]
```

Comment:
This is very illuminating.  I skipped over it at first and then wondered "Why all the code duplication?". I'm not going to claim I understand all of the details though.  @effectfully You're good at persuading GHC to do what you want: do you have any comments/suggestions?

Reply:
I would also love to know if this can be written with less duplicative code. I tried about every trick I know, but anything using HOFs seemed to run like treacle.

Reply:
@effectfully says:

I feel like helper1 would work with an explicit inline call. Is it worth it though?

so like this:
>   While in theory inlining could help with at least the `helper1` case (`helper2` is beyond that technique), it doesn't seem like GHC is able to figure this out, even with `INLINE` is placed on `helper1`.

but also make it `inline helper1` at the call site.

Maybe it's also possible to turn `helper1` into a type class with a default implementation that doesn't take the functions as arguments but instead uses another method of the type class added specifically so that GHC knows what it is.
Or helper1 could take an argument `whichOp` and match on it inside the body of `helper1` to pick either `.&.` or `.|.` or ... One extra pattern match that should be fairly cheap and no headache with GHC not inlining things.

Reply:
I tried an explicit `INLINE` and it didn't help at all.

Reply:
`INLINE` is not enough in some cases, you may need to `GHC.Magic.inline` at the call site. I'd be surprised if it didn't make the call inlined, `inline` is pretty reliable (unlike `INLINE`).